File: sdks/go/jws/ed25519.go
package jws

import (
	"crypto/ed25519"
	"fmt"
	"math/big"
)

// Ed25519PublicKey represents an Ed25519 public key.
type Ed25519PublicKey struct {
	Key ed25519.PublicKey
	KID string
}

// ed25519GroupOrderL is the Ed25519 group order L = 2^252 +
// 27742317777372353535851937790883648493. The signature scalar S (signature
// bytes 32..64, little-endian) must be reduced modulo L; S >= L is non-canonical
// and is rejected (RFC 8032 malleability guard).
var ed25519GroupOrderL, _ = new(big.Int).SetString(
	"1000000000000000000000000000000014def9dea2f79cd65812631a5cf5d3ed", 16)

// scalarFromSignature reads the signature scalar S (bytes 32..64,
// little-endian) as a big.Int.
func scalarFromSignature(signature []byte) *big.Int {
	le := signature[32:64]
	be := make([]byte, len(le))
	for i := range le {
		be[i] = le[len(le)-1-i]
	}
	return new(big.Int).SetBytes(be)
}

// VerifyEd25519 verifies an Ed25519 signature under the PEAC Ed25519
// verification profile.
//
// Profile = cofactorless Ed25519 verification plus bounded admissibility checks
// over the public inputs, applied in order:
//  1. length: public key is 32 bytes, signature is 64 bytes;
//  2. bounded admissibility precheck on the public key A;
//  3. reject non-reduced scalars S >= L (RFC 8032 malleability guard);
//  4. bounded admissibility precheck on the signature component R;
//  5. cofactorless verification via crypto/ed25519.Verify.
//
// "RFC 8032 strict" is not a single predicate: libraries differ on small-order
// points and cofactored-versus-cofactorless verification. Go's stdlib is
// cofactorless; the TypeScript reference verifier (packages/crypto/src/ed25519.ts)
// uses Web Crypto, which is also cofactorless, with the identical admissibility
// checks above, so both implementations accept and reject the same signatures
// across the shared edge-vector corpus
// (specs/conformance/parity-corpus/ed25519-peac-profile/).
func VerifyEd25519(publicKey ed25519.PublicKey, message, signature []byte) error {
	if len(publicKey) != ed25519.PublicKeySize {
		return fmt.Errorf("invalid public key size: expected %d, got %d", ed25519.PublicKeySize, len(publicKey))
	}

	if len(signature) != ed25519.SignatureSize {
		return fmt.Errorf("invalid signature size: expected %d, got %d", ed25519.SignatureSize, len(signature))
	}

	if reason := ed25519RejectionReason(ed25519PointFrom(publicKey)); reason != ed25519NoRejection {
		return fmt.Errorf("inadmissible public key encoding rejected: %s", reason)
	}

	if scalarFromSignature(signature).Cmp(ed25519GroupOrderL) >= 0 {
		return fmt.Errorf("non-canonical signature scalar (S >= L) rejected")
	}

	// The same bounded precheck on the signature component R. Reached only after the 64-byte
	// length check above, so this slice is always in range.
	if reason := ed25519RejectionReason(ed25519PointFrom(signature[:32])); reason != ed25519NoRejection {
		return fmt.Errorf("inadmissible signature R encoding rejected: %s", reason)
	}

	if !ed25519.Verify(publicKey, message, signature) {
		return fmt.Errorf("signature verification failed")
	}

	return nil
}

// VerifyJWS verifies a JWS using Ed25519.
func VerifyJWS(jws *ParsedJWS, publicKey ed25519.PublicKey) error {
	if jws.Header.Algorithm != "EdDSA" {
		return fmt.Errorf("unsupported algorithm: %s", jws.Header.Algorithm)
	}

	return VerifyEd25519(publicKey, jws.SigningInput, jws.Signature)
}

// ParsePublicKeyFromBytes parses an Ed25519 public key from raw bytes.
func ParsePublicKeyFromBytes(data []byte) (ed25519.PublicKey, error) {
	if len(data) != ed25519.PublicKeySize {
		return nil, fmt.Errorf("invalid public key size: expected %d, got %d", ed25519.PublicKeySize, len(data))
	}
	return ed25519.PublicKey(data), nil
}

File: packages/crypto/src/ed25519.ts
/**
 * Internal Ed25519 wrapper -- async-only surface
 *
 * Signing uses the async methods from @noble/ed25519 (Web Crypto backed).
 * In noble v3, sync methods require explicit hash configuration; async methods
 * use built-in Web Crypto and need no configuration.
 *
 * Verification uses the PEAC Ed25519 verification profile (see verify() below),
 * which is implemented on the Web Crypto Subtle API rather than noble, because
 * noble's verifier (both ZIP215 default and { zip215: false }) uses cofactored
 * verification on certain edge inputs and therefore does NOT agree with the Go
 * reference verifier (crypto/ed25519, cofactorless) on the ed25519-speccheck
 * corpus. Web Crypto Ed25519 is cofactorless and matches Go.
 *
 * This module re-exports only the async surface to prevent accidental sync
 * usage. All other modules in @peac/crypto MUST import from this file, never
 * directly from '@noble/ed25519'.
 *
 * Key material handling:
 * - Private keys are 32-byte Uint8Array (Ed25519 seed)
 * - Public keys are 32-byte Uint8Array (compressed Ed25519 point)
 * - JavaScript cannot guarantee memory zeroization; callers should avoid
 *   storing keys longer than necessary and never log key bytes
 * - randomSecretKey() uses crypto.getRandomValues() (CSPRNG)
 */

// Namespace import avoids tsup tree-shaking false positive in multi-entry builds
// where signAsync appears unused in the testkit entry point.
import * as ed from '@noble/ed25519';

import { isRejectedEd25519PointEncoding } from './internal/ed25519-admissibility';

/** Sign a message with Ed25519 (async, Web Crypto backed) */
export const sign = ed.signAsync;

/** Derive public key from private key (async, Web Crypto backed) */
export const getPublicKey = ed.getPublicKeyAsync;

/** Generate a cryptographically random 32-byte secret key (CSPRNG) */
export const randomSecretKey = ed.utils.randomSecretKey;

const ED25519_PUBLIC_KEY_BYTES = 32;
const ED25519_SIGNATURE_BYTES = 64;

/**
 * Ed25519 group order L = 2^252 + 27742317777372353535851937790883648493,
 * encoded as 32 little-endian bytes (the same layout as the signature scalar S,
 * which occupies signature bytes 32..64 little-endian). The scalar S MUST be
 * reduced modulo L; values S >= L are non-canonical and are rejected (the
 * RFC 8032 malleability guard, part of the PEAC Ed25519 verification profile).
 * Stored as bytes (not a bigint) so the admissibility check is a fixed-width
 * byte comparison with no allocation.
 */
const ED25519_GROUP_ORDER_L_LE = Uint8Array.from([
  0xed, 0xd3, 0xf5, 0x5c, 0x1a, 0x63, 0x12, 0x58, 0xd6, 0x9c, 0xf7, 0xa2, 0xde, 0xf9, 0xde, 0x14,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x10,
]);

/**
 * True if the signature scalar S (signature bytes 32..64, little-endian) is
 * greater than or equal to the Ed25519 group order L. Both S and L are 32-byte
 * little-endian magnitudes; compare from the most significant byte (index 31)
 * down. No bigint, no allocation; constant 32-byte work over public inputs.
 */
function scalarIsNonCanonical(signature: Uint8Array): boolean {
  for (let i = ED25519_PUBLIC_KEY_BYTES - 1; i >= 0; i--) {
    const sByte = signature[ED25519_PUBLIC_KEY_BYTES + i];
    const lByte = ED25519_GROUP_ORDER_L_LE[i];
    if (sByte !== lByte) return sByte > lByte;
  }
  // S == L exactly: non-canonical (S must be strictly < L).
  return true;
}

/**
 * Thrown only when the runtime cannot provide the cofactorless Ed25519
 * primitive PEAC requires. The verifier fails closed: it never silently falls
 * back to a different verification predicate.
 */
export class Ed25519RuntimeError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'Ed25519RuntimeError';
  }
}

// Web Crypto types (CryptoKey, BufferSource, SubtleCrypto) are not in this
// package's `lib` ("ES2022", no DOM). Derive what we need from the ambient
// `globalThis.crypto` typing provided by @types/node, mirroring the pattern in
// hash.ts and @peac/http-signatures, so no DOM lib is required.
type Subtle = NonNullable<typeof globalThis.crypto>['subtle'];

function subtleCrypto(): Subtle {
  const c = globalThis.crypto;
  if (!c || !c.subtle) {
    throw new Ed25519RuntimeError(
      'Web Crypto SubtleCrypto is unavailable; PEAC Ed25519 verification requires it'
    );
  }
  return c.subtle;
}

/**
 * Read an error's `name` structurally (no `instanceof DOMException`, which is
 * not defined in every runtime). Used to tell an unsupported-algorithm runtime
 * failure apart from a malformed-input rejection.
 */
function errorName(err: unknown): string | undefined {
  return err !== null &&
    typeof err === 'object' &&
    'name' in err &&
    typeof (err as { name: unknown }).name === 'string'
    ? (err as { name: string }).name
    : undefined;
}

/**
 * True when a Web Crypto operation failed because the runtime does not support
 * the Ed25519 algorithm (as opposed to rejecting a malformed/invalid input).
 * Web Crypto raises NotSupportedError for an unsupported algorithm. We map this
 * to a fail-closed Ed25519RuntimeError so an unsupported runtime is never
 * misreported as an invalid signature.
 */
function isUnsupportedEd25519Runtime(err: unknown): boolean {
  return errorName(err) === 'NotSupportedError';
}

/**
 * Verify an Ed25519 signature under the PEAC Ed25519 verification profile.
 *
 * Profile order, identical in the Go reference verifier:
 *   1. validate public-key and signature lengths;
 *   2. bounded admissibility precheck on the public key A;
 *   3. reject a non-reduced scalar S >= L (the RFC 8032 malleability guard);
 *   4. bounded admissibility precheck on the signature component R;
 *   5. delegate remaining curve validity and the signature equation to the runtime primitive.
 *
 * Steps 2 and 4 are the SAME bounded precheck applied to two different inputs. It rejects encoded
 * y >= p, the two x = 0 encodings whose sign bit is set, the eight canonical torsion encodings and
 * two canonical order-4L encodings retained by PEAC policy. It performs neither complete point
 * decoding, complete curve validation, prime-subgroup membership testing, nor complete mixed-order
 * rejection.
 *
 * Rationale: runtimes disagree at these edges. Web Cryptography Level 2 rejects invalid or
 * small-order A and R before applying the cofactorless equation and notes implementations have not
 * behaved uniformly; RFC 8032 permits the cofactored equation, whose accepted set differs here.
 * Deciding the enumerated cases in PEAC makes the outcome independent of the primitive underneath.
 *
 * Verify-only: signing output is unchanged, so every signature produced by sign() remains valid.
 *
 * Fails closed: if the runtime lacks Web Crypto Ed25519 this throws Ed25519RuntimeError. It never
 * falls back to a different verification path, which would silently change the predicate.
 */
export async function verify(
  signature: Uint8Array,
  message: Uint8Array,
  publicKey: Uint8Array
): Promise<boolean> {
  // 1. Length checks (constant, public inputs).
  if (publicKey.length !== ED25519_PUBLIC_KEY_BYTES) return false;
  if (signature.length !== ED25519_SIGNATURE_BYTES) return false;

  // 2. Bounded admissibility precheck on the public key.
  if (isRejectedEd25519PointEncoding(publicKey)) return false;

  // 3. Reject non-reduced scalar S >= L (fixed-width little-endian byte
  // comparison; both operands are public signature bytes, an admissibility
  // guard, not secret-dependent key handling).
  if (scalarIsNonCanonical(signature)) return false;

  // 4. The same bounded precheck on the signature component R. Reached only after the 64-byte
  //    length check above, so this zero-copy view is always in range.
  if (isRejectedEd25519PointEncoding(signature.subarray(0, ED25519_PUBLIC_KEY_BYTES))) return false;

  // 5. Cofactorless verification via Web Crypto (fails closed if unavailable).
  const subtle = subtleCrypto();
  let key: Awaited<ReturnType<Subtle['importKey']>>;
  try {
    key = await subtle.importKey('raw', publicKey, { name: 'Ed25519' }, false, ['verify']);
  } catch (err) {
    // A runtime that has crypto.subtle but does NOT support Ed25519 raises
    // NotSupportedError here. That is a runtime-capability failure, not an
    // invalid key: fail closed rather than misreport it as an invalid
    // signature. A malformed / non-decodable public-key encoding (DataError or
    // similar) is a genuine rejection and returns false.
    if (isUnsupportedEd25519Runtime(err)) {
      throw new Ed25519RuntimeError(
        'Web Crypto Ed25519 verification is unavailable in this runtime'
      );
    }
    return false;
  }
  try {
    return await subtle.verify({ name: 'Ed25519' }, key, signature, message);
  } catch (err) {
    // Same distinction at the verify step: an unsupported-algorithm runtime
    // fails closed; any other operational error is a verification failure.
    if (isUnsupportedEd25519Runtime(err)) {
      throw new Ed25519RuntimeError(
        'Web Crypto Ed25519 verification is unavailable in this runtime'
      );
    }
    return false;
  }
}
